### PR TITLE
ZMI: limit number of resulted folderish objects and show SVG as FSImage

### PR DESCRIPTION
**Pull request by drfho**
The manage_main-vIew of FSDirectoryVIew object may extremly slow down the system if the DirectoryView is next to root and the ZODB has a lot of objects. Reason is that all (!) folderish Zope objects are listed in the select-list "customize" (for selecting a copy-target).
Actually the manage_main-view ist primarily used for getting a short glance of the object's code/content (and maybe not so often for replicating its code into a ZODB object).
So I would like to suggest a max-parameter to ensure a quick response of the manage_main view. I propose a default value of 20 that can be adjusted to (50,100.200.500.1000, unlimited) on click, if the long list is really needed. 

[1] Former ZMI (master branch)
zmi_cusomize_item0

[2] New ZMI status
zmi_cusomize_item3

### CODE REVIEW

File: src/Products/CMFCore/DirectoryView.py [hunk 1]
```diff
@@ -359,10 +359,11 @@ def listDirectories(self):
 registerMetaType = _dirreg.registerMetaType
 
 
-def listFolderHierarchy(ob, path, rval, adding_meta_type=None):
+def listFolderHierarchy(ob, path, rval, adding_meta_type=None, max=0, c=0):
```
Comment (dataflake) [inline src/Products/CMFCore/DirectoryView.py:362]: Please use meaningful parameter names, not just `c`

File: src/Products/CMFCore/dtml/custdtml.dtml [hunk 2]
```diff
@@ -30,17 +30,41 @@
...
+          </div>
+          <div cass="input-group-append">
+            <input class="btn btn-primary" type="submit" name="submit" value=" Customize " />
+          </div>
+        </div>
+        <small class="text-muted">
           Select a destination folder and press the button
           to make a copy of this method that can be customized.
+          If the destination folder is not listed, then 
+          enlarge the amount of maximum results (default: 20).
```
Comment (dataflake) [inline src/Products/CMFCore/dtml/custdtml.dtml:67]:
```suggestion
          If the destination folder is not listed try 
          increasing the maximum results number (default: 20).
```

File: src/Products/CMFCore/dtml/custfile.dtml [hunk 3]
```diff
@@ -48,17 +48,41 @@
...
+          </div>
+          <div cass="input-group-append">
+            <input class="btn btn-primary" type="submit" name="submit" value=" Customize " />
+          </div>
+        </div>
+        <small class="text-muted">
           Select a destination folder and press the button
           to make a copy of this method that can be customized.
+          If the destination folder is not listed, then 
+          enlarge the amount of maximum results (default: 20).
```
Comment (dataflake) [inline src/Products/CMFCore/dtml/custfile.dtml:85]:
```suggestion
          If the destination folder is not listed try 
          increasing the maximum results number (default: 20).
```

File: src/Products/CMFCore/dtml/custimage.dtml [hunk 4]
```diff
@@ -57,17 +57,41 @@
...
+          </div>
+          <div cass="input-group-append">
+            <input class="btn btn-primary" type="submit" name="submit" value=" Customize " />
+          </div>
+        </div>
+        <small class="text-muted">
           Select a destination folder and press the button
           to make a copy of this method that can be customized.
+          If the destination folder is not listed, then 
+          enlarge the amount of maximum results (default: 20).
```
Comment (dataflake) [inline src/Products/CMFCore/dtml/custimage.dtml:94]:
```suggestion
          If the destination folder is not listed try 
          increasing the maximum results number (default: 20).
```

File: src/Products/CMFCore/dtml/custpt.dtml [hunk 5]
```diff
@@ -30,17 +30,41 @@
...
+          </div>
+          <div cass="input-group-append">
+            <input class="btn btn-primary" type="submit" name="submit" value=" Customize " />
+          </div>
+        </div>
+        <small class="text-muted">
           Select a destination folder and press the button
           to make a copy of this method that can be customized.
+          If the destination folder is not listed, then 
+          enlarge the amount of maximum results (default: 20).
```
Comment (dataflake) [inline src/Products/CMFCore/dtml/custpt.dtml:67]:
```suggestion
          If the destination folder is not listed try 
          increasing the maximum results number (default: 20).
```

File: src/Products/CMFCore/dtml/custpy.dtml [hunk 6]
```diff
@@ -30,17 +30,41 @@
...
+          </div>
+          <div cass="input-group-append">
+            <input class="btn btn-primary" type="submit" name="submit" value=" Customize " />
+          </div>
+        </div>
+        <small class="text-muted">
           Select a destination folder and press the button
           to make a copy of this method that can be customized.
+          If the destination folder is not listed, then 
+          enlarge the amount of maximum results (default: 20).
```
Comment (dataflake) [inline src/Products/CMFCore/dtml/custpy.dtml:67]:
```suggestion
          If the destination folder is not listed try 
          increasing the maximum results number (default: 20).
```

File: src/Products/CMFCore/dtml/custzsql.dtml [hunk 7]
```diff
@@ -30,17 +30,41 @@
...
+          </div>
+          <div cass="input-group-append">
+            <input class="btn btn-primary" type="submit" name="submit" value=" Customize " />
+          </div>
+        </div>
+        <small class="text-muted">
           Select a destination folder and press the button
           to make a copy of this method that can be customized.
+          If the destination folder is not listed, then 
+          enlarge the amount of maximum results (default: 20).
```
Comment (dataflake) [inline src/Products/CMFCore/dtml/custzsql.dtml:67]:
```suggestion
          If the destination folder is not listed try 
          increasing the maximum results number (default: 20).
```

**Review by dataflake — CHANGES_REQUESTED** [review]

File: src/Products/CMFCore/DirectoryView.py (hunk 1 above)
Comment (drfho) [reply src/Products/CMFCore/DirectoryView.py]: Thanks for the hint. Done: https://github.com/zopefoundation/Products.CMFCore/pull/130/commits/823ba67709f6391911f51adc315e9a275e371861

**Review by drfho — COMMENTED** [review]
zmi_cusomize_item4_stx

**Comment by drfho** [comment]
Hi @dataflake ,
now my changes are considering Your kind hints. I hope the ZMI improvement meets Your expectation and the PR is ready to merge.
Many thanks 
f

**Review by dataflake — APPROVED** [review]

**Comment by dataflake** [comment]
Sorry for being nitpicky. Your cleanup work is always appreciated.

**Review by drfho — COMMENTED** [review]
Sorry, count-param was not needed; it works easier with len() of the rval-list.

**Comment by dataflake** [comment]
👍

**Review by drfho — COMMENTED** [review]
SVG files are considered as images now:

svg_image_fsview

**Comment by dataflake** [comment]
Just make sure that this switch from FSFile to FSImage does not lead to blowups/tracebacks when used with Zope versions from before https://github.com/zopefoundation/Zope/pull/1146

**Review by drfho — COMMENTED** [review]
Thanks for the hint (https://github.com/zopefoundation/Products.CMFCore/pull/130#issuecomment-1694617804); actually with default width/height=-1 the FSImage-preview does not work (same as FSFile).

This  _custimage_-DTML-template change provides an image preview now, even without zopefoundation/Zope#1146.
Having _no image dimensions_ the SVG pic is shown using the `object`-Tag for setting mime-type (!)  and an onload-resizing for a best-guess-fit. This is not as perfect as scaling with available dimensions (zopefoundation/Zope#1146) , but it is still an UX-improvement compared to the former status of SVG as FSFile, because formerly the image-preview was not shown at all.

_LEFT: without Zope-PR#1146; RIGHT: with Zope-PR#1146_
FSImage_backwards

**Comment by dataflake** [comment]
👍

**Review by drfho — COMMENTED** [review]
White or transparent SVG images cannot be seen on a white background. Photoshop et.al. use a checkerboard pattern in case:

checkerboard

**Review by drfho — COMMENTED** [review]
More usertests revealed that for white+transparent images the contrasts may be improved; now onhover the checkerboard resolves into plain grey for a better visibbility of the details.

svg_trans_hover

**Comment by drfho** [comment]
Hi @dataflake , 
now the PR may be mature for merging, My or your turn?
Mabe in combination with https://github.com/zopefoundation/Zope/pull/1146
Many thanks
f

**Comment by dataflake** [comment]
The PR author merges after approval, which you already have for both. Next time I would suggest you ask for approval at the very end. I didn't know this was going to hang around for such a long time while you were making tweaks and fixes.

**Pull request merged**